// zustand/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var react = require('react');

var useIsoLayoutEffect = typeof window === 'undefined' ? react.useEffect : react.useLayoutEffect;
function create(createState) {
  var state;
  var listeners = new Set();

  var setState = function setState(partial) {
    var partialState = typeof partial === 'function' ? partial(state) : partial;

    if (partialState !== state) {
      state = Object.assign({}, state, partialState);
      listeners.forEach(function (listener) {
        return listener();
      });
    }
  };

  var getState = function getState() {
    return state;
  };

  var getSubscriber = function getSubscriber(listener, selector, equalityFn) {
    if (selector === void 0) {
      selector = getState;
    }

    if (equalityFn === void 0) {
      equalityFn = Object.is;
    }

    return {
      currentSlice: selector(state),
      equalityFn: equalityFn,
      errored: false,
      listener: listener,
      selector: selector,
      unsubscribe: function unsubscribe() {}
    };
  };

  var subscribe = function subscribe(subscriber) {
    function listener() {
      // Selector or equality function could throw but we don't want to stop
      // the listener from being called.
      // https://github.com/react-spring/zustand/pull/37
      try {
        var newStateSlice = subscriber.selector(state);

        if (!subscriber.equalityFn(subscriber.currentSlice, newStateSlice)) {
          subscriber.listener(subscriber.currentSlice = newStateSlice);
        }
      } catch (error) {
        subscriber.errored = true;
        subscriber.listener(null, error);
      }
    }

    listeners.add(listener);
    return function () {
      listeners.delete(listener);
    };
  };

  var apiSubscribe = function apiSubscribe(listener, selector, equalityFn) {
    return subscribe(getSubscriber(listener, selector, equalityFn));
  };

  var destroy = function destroy() {
    return listeners.clear();
  };

  var useStore = function useStore(selector, equalityFn) {
    if (selector === void 0) {
      selector = getState;
    }

    if (equalityFn === void 0) {
      equalityFn = Object.is;
    }

    var forceUpdate = react.useReducer(function (c) {
      return c + 1;
    }, 0)[1];
    var subscriberRef = react.useRef();

    if (!subscriberRef.current) {
      subscriberRef.current = getSubscriber(forceUpdate, selector, equalityFn);
      subscriberRef.current.unsubscribe = subscribe(subscriberRef.current);
    }

    var subscriber = subscriberRef.current;
    var newStateSlice;
    var hasNewStateSlice = false; // The selector or equalityFn need to be called during the render phase if
    // they change. We also want legitimate errors to be visible so we re-run
    // them if they errored in the subscriber.

    if (subscriber.selector !== selector || subscriber.equalityFn !== equalityFn || subscriber.errored) {
      // Using local variables to avoid mutations in the render phase.
      newStateSlice = selector(state);
      hasNewStateSlice = !equalityFn(subscriber.currentSlice, newStateSlice);
    } // Syncing changes in useEffect.


    useIsoLayoutEffect(function () {
      if (hasNewStateSlice) {
        subscriber.currentSlice = newStateSlice;
      }

      subscriber.selector = selector;
      subscriber.equalityFn = equalityFn;
      subscriber.errored = false;
    });
    useIsoLayoutEffect(function () {
      return subscriber.unsubscribe;
    }, []);
    return hasNewStateSlice ? newStateSlice : subscriber.currentSlice;
  };

  var api = {
    setState: setState,
    getState: getState,
    subscribe: apiSubscribe,
    destroy: destroy
  };
  state = createState(setState, getState, api);
  return [useStore, api];
}

exports.create = create;
exports.default = create;
